### PR TITLE
debug: state-inspect debugging utility

### PR DESCRIPTION
**Pull request by stolowski**
Implement new snap debug commands for inspecting state.json without involving snapd, so that some aspects of state can be examined without needing to load it through snapd (which inevitably modifies it and the outcome, e.g. some changes might get executed/undone) - useful to debug issues from users. It already proved to be useful for debugging one such issue.

This is just a starting point, there are probably more aspects of change, tasks and other state objects that can be added later as needs arise - for example a way to inspect connections ("conns") with all their internal attributes.

Example usage:
```
snap debug tasks --change-id=203 /var/lib/snapd/state.json
snap debug tasks --change-id=203     # state path can be omitted, defaults to state.json in CWD
snap debug tasks --change-id=203 --no-hold
snap debug changes
```

### CODE REVIEW

**Comment by stolowski** [comment]
> Hey Pawel
> 
> I like the general idea. This will give us some ability to both interactively look at things without having to jump through hoops with `jq`. It also buys us some future proofing for eventual format change.
> 
> Having said that some reservations:
> 
> * should we only add the commands that are implemented now, add remaining commands as they are implemented?
> * should both snapd and the debug utility use flock based lock for mutual exclusion?
> * once approved, could you do a quick pass with @degville on command help strings?
> * could you add a super-basic spread test that just runs the existing commands and checks that something sensible got printed?

Hey, Thanks for the remarks!

It seems that per yesterday's discussion on the standup the general direction with this tool will be to integrate it with `snap debug ....`. With that it will definately get some tests, including a spread test, and I will check with Graham and tweak help strings; also in this caseI agree  it makes sense to drop unimplemented commands. 

Locking is an interesting idea, although given the very specific audience and goal of this tool (inspecting an "offline" state, not the one in use) I'm not sure it's actually necessary (also, I'd keep it unintrusive and avoid introducing state locking in snapd as it normally doesn't need it). I'm open to discuss that.

**Comment by stolowski** [comment]
I'll return to this PR in a while, once more important PRs are dealt with.

**Comment by stolowski** [comment]
I've re-worked this PR and moved the new commands under `snap debug ...`. Also added a spread test and some unit tests.

File: cmd/snap/cmd_debug_changes.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+func (c byChangeID) Swap(i, j int)      { c[i], c[j] = c[j], c[i] }
+func (c byChangeID) Less(i, j int) bool { return c[i].ID() < c[j].ID() }
+
+func formatTime(t time.Time) string {
+	return t.Format(time.RFC3339)
+}
+
+func loadState(path string) (*state.State, error) {
+	if path == "" {
+		path = "state.json"
```
Comment (bboozzoo) [inline cmd/snap/cmd_debug_changes.go]: Should this case fall back to open the actual `/var/lib/snapd/state.json` file?

File: cmd/snap/cmd_debug_task.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	if termWidth > 100 {
+		// any wider than this and it gets hard to read
+		termWidth = 100
+	}
+
+	// the output of 'debug task' is yaml'ish
+	fmt.Fprintf(Stdout, "id: %s\nkind: %s\nsummary: %s\nstatus: %s\n\n", taskID, task.Kind(), task.Summary(), task.Status().String())
+	log := task.Log()
+	if len(log) > 0 {
+		fmt.Fprintf(Stdout, "log:\n")
```
Comment (bboozzoo) [inline cmd/snap/cmd_debug_task.go]: This should probably be `log: |\n` to be yamlish

File: cmd/snap/cmd_debug_task.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+			if err := wrapLine(Stdout, []rune(msg), "  ", termWidth); err != nil {
+				break
+			}
+		}
+		fmt.Fprintln(Stdout)
+	}
+
+	fmt.Fprintf(Stdout, "tasks waiting for %s:\n", taskID)
+	for _, ht := range task.HaltTasks() {
+		fmt.Fprintf(Stdout, "  %s (%s)\n", ht.Kind(), ht.ID())
```
Comment (bboozzoo) [inline cmd/snap/cmd_debug_task.go]: `"  - %s (%s)"` then?

File: tests/main/snap-debug-tasks-changes/task.yaml [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+    echo "Changes can be listed"
+    snap debug changes /var/lib/snapd/state.json | MATCH "seed .*Initialize system state"
+
+    echo "Snap changes defaults to state.json in the current directory"
+    cd /var/lib/snapd
+    snap debug changes | MATCH "install-snap .*Install \"hello-world\" snap"
+
+    echo "Tasks can be listed"
+    snap debug tasks --change-id=1 /var/lib/snapd/state.json | MATCH "mark-seeded .*Mark system seeded"
```
Comment (bboozzoo) [inline tests/main/snap-debug-tasks-changes/task.yaml]: Could use a test with `snap debug tasks --change-id=1 --dot /var/lib/snapd/state.json | dot -Tpng > out.png`

File: cmd/snap/cmd_debug_changes.go (hunk 1 above)
Comment (stolowski) [reply cmd/snap/cmd_debug_changes.go]: I'm on the fence with this, it depends on typical use case. The driver for this was to debug someone else's state.json, meaning I'm usually interested in state.json from my CWD.

**Comment by pedronis** [comment]
I don't have a counterproposal but I find using the same names as the main (non-debug) commands but with slighly different syntax confusing.

**Comment by stolowski** [comment]
> I don't have a counterproposal but I find using the same names as the main (non-debug) commands but with slighly different syntax confusing.

How about `snap debug state-changes ...` (and `state-task`)?

**Comment by pedronis** [comment]
my current preference is too try not to have too many commands but do something like:

snap debug state [--changes|--task=TID|--change=CHGID]  [state.json]

this let us later add option to explore other parts of state if needed.

snap debug state

without any options would be the same as --changes for now.

**Comment by stolowski** [comment]
Re-worked as `snap debug state ....`, per Samuele's suggestions.

File: cmd/snap/cmd_debug_state_test.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+	"path/filepath"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/cmd/snap"
+)
+
+var stateJSON = []byte(`
+{
+	"last-task-id": 999,
```
Comment (anonymouse64) [inline cmd/snap/cmd_debug_state_test.go]: unclear if this matters, but isn't this an incorrect task-id? I don't see 999 anywhere in the changes's tasks

File: cmd/snap/cmd_debug_state_test.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+	"last-change-id": 99,
```
Comment (anonymouse64) [inline cmd/snap/cmd_debug_state_test.go]: same comment about change ID, I don't see a change 99?

File: cmd/snap/cmd_debug_state_test.go (hunk 5 above)
Comment (stolowski) [reply cmd/snap/cmd_debug_state_test.go]: Indeed, the test state is not consistient, I'll fix that, thanks.

File: cmd/snap/cmd_debug_state.go [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+var cmdDebugStateShortHelp = i18n.G("Inspect a snapd state file.")
+var cmdDebugStateLongHelp = i18n.G("Inspect a snapd state file, bypassing snapd API.")
+
+type byChangeID []*state.Change
+
+func (c byChangeID) Len() int           { return len(c) }
+func (c byChangeID) Swap(i, j int)      { c[i], c[j] = c[j], c[i] }
+func (c byChangeID) Less(i, j int) bool { return c[i].ID() < c[j].ID() }
+
+func formatTime(t time.Time) string {
```
[the remark this answers is not in the document]
Comment (stolowski) [reply cmd/snap/cmd_debug_state.go]: Nice finding, I didn't know of timeMixin! Done.

File: cmd/snap/cmd_debug_state_test.go (hunk 6 above)
Comment (stolowski) [reply cmd/snap/cmd_debug_state_test.go]: Updated both, although this is purely cosmetic, these numbers are only important for state when creating new tasks & allocating IDs for them.

File: cmd/snap/cmd_debug_state.go [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+
+	termWidth, _ := termSize()
+	termWidth -= 3
+	if termWidth > 100 {
+		// any wider than this and it gets hard to read
+		termWidth = 100
+	}
+
+	// the output of 'debug task' is yaml'ish
+	fmt.Fprintf(Stdout, "id: %s\nkind: %s\nsummary: %s\nstatus: %s\n\n", taskID, task.Kind(), task.Summary(), task.Status().String())
```
[the remark this answers is not in the document]
Comment (stolowski) [reply cmd/snap/cmd_debug_state.go]: Sorry, I forgot to reply to that question: I don't think we have such cases right now, we try to keep them brief as they show up in other contexts (e.g. `snap changes`) in tabular output, so I woudln't worry about that.

**Pull request merged**